### ISSUE TITLE
PHPToExcel::timestampToExcel - Forces GMT timezone

### ISSUE THREAD

**Issue by fyrye** [PHPOffice/PhpSpreadsheet#611]
This is:
- [x] a bug report
- [ ] a feature request
- [X] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)

### What is the expected behavior?
When supplying  a unix tImestamp it should be displayed in the specified application timezone and function identically to the `dateTimeToExcel` and `stringToExcel` functions.

When specifying a unix timestamp (`strtotime('2018-01-01 13:00:00')`),  the  Excel date should be represented as  `2018-01-01 13:00:00`. Identically to using a DateTime object (`\DateTime('2018-01-01 13:00:00')`)

### What is the current behavior?
`PHPToExcel::timestampToExcel` uses the `new \DateTime('@' . $timestamp)` syntax. Which ignores the specified application timezone, instead using the default GMT timezone, and produces unexpected results, where `dateTimeToExcel` uses the specified DateTime object timezone 

The represented Excel date is `2018-01-01 18:00:00`
See PHP example:  https://3v4l.org/H50W7

The issue is compounded when an application specifies a timezone that observes Daylight Savings Time (DST), where the represented GMT time is displayed with the adjusted hour.
Example: https://3v4l.org/mhIqn

### What are the steps to reproduce?

```php
<?php
require_once __DIR__ . '/vendor/autoload.php';

// Create new Spreadsheet object
$spreadsheet = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

// add code that show the issue here...
date_default_timezone_set('America/New_York');
$dateString = '2018-01-01 13:00:00';
$unixTimestamp = strtotime($dateString);
$dateObject = new \DateTime($dateString);

#draw the specified date time to excel
$worksheet = $spreadsheet->setActiveSheetIndex(0);

#displays 2018-01-01 13:00:00
$worksheet
    ->getCellByColumnAndRow(1, 1)
    ->setValue(\PhpOffice\PhpSpreadsheet\Shared\Date::PHPToExcel($dateString))
    ->getStyle()
    ->getNumberFormat()
    ->setFormatCode(\PhpOffice\PhpSpreadsheet\Style\NumberFormat::FORMAT_DATE_DATETIME);

#displays: 2018-01-01 13:00:00
$worksheet
    ->getCellByColumnAndRow(1, 2)
    ->setValue(\PhpOffice\PhpSpreadsheet\Shared\Date::PHPToExcel($dateObject))
    ->getStyle()
    ->getNumberFormat()
    ->setFormatCode(\PhpOffice\PhpSpreadsheet\Style\NumberFormat::FORMAT_DATE_DATETIME);

#displays: 2018-01-01 18:00:00
$worksheet
    ->getCellByColumnAndRow(1, 3)
    ->setValue(\PhpOffice\PhpSpreadsheet\Shared\Date::PHPToExcel($unixTimestamp))
    ->getStyle()
    ->getNumberFormat()
    ->setFormatCode(\PhpOffice\PhpSpreadsheet\Style\NumberFormat::FORMAT_DATE_DATETIME);

```

### Which versions of PhpSpreadsheet and PHP are affected?

- PHP `5.6`
- PhpSpreadsheet `1.3.1`

### Suggested Resolutions 
_Goal: reproduce same results as `dateTimeToExcel` and `stringToExcel`_

1) Use `setTimestamp` instead of `@` - correct usage _(BC Break)_
```php
    public static function timestampToExcel($dateValue)
    {
        if (!is_numeric($dateValue)) {
            return false;
        }
        $dateObject = new \DateTime;

        return self::dateTimeToExcel($dateObject->setTimeStamp($dateValue));
    }
```
2) Use `self::stringToExcel` instead of `self::dateTimeToExcel` _(BC Break)_

```php

    public static function timestampToExcel($dateValue)
    {
        if (!is_numeric($dateValue)) {
            return false;
        }

        return self::stringToExcel(date('Y-m-d H:i:s', $dateValue));
    }
```
3) Add desired timezone null, text or object argument, default to `UTC`. _(BC compliant)_
_`null` uses specified application timezone._
Example: https://3v4l.org/l2jjV

```php
    public static function timestampToExcel($dateValue, $tz = 'UTC')
    {
        if (!is_numeric($dateValue)) {
            return false;
        }
        if (null !== $tz) {
             if (is_string($tz)) {
                $tz = new \DateTimeZone($tz);
             }
        }
        
        $dateObject = new \DateTime('', $tz);
       
        return self::dateTimeToExcel($dateObject->setTimestamp($dateValue));
    }
```

4) Add desired timezone null, text or object argument default to `null`. _(BC compliant)_
_`null` ignores specified application timezone_
```php
    public static function timestampToExcel($dateValue, $tz = null)
    {
        if (!is_numeric($dateValue)) {
            return false;
        }
        
         $dateObject = new \DateTime('@' . $dateValue);
         if (null !== $tz) {
             if (is_string($tz)) {
                $tz = new \DateTimeZone($tz);
             }
             $dateObject->setTimezone($tz);
        }
        return self::dateTimeToExcel($dateObject);
    }
```

If the forced use of a GMT timezone is expected behavior, it should be documented on the `timestampToExcel` and `PHPToExcel` methods, that unix timestamps will use a GMT timezone and not the specified application timezone.

### TLDR:
Albeit it is best practice to utilize `UTC` as the default timezone for an application. When a developer specifies a timezone to be used throughout the application, the behavior should not be altered by third-party libraries. Using `new DateTime('@' . $timestamp)` ignores the specified application timezone, producing an unexpected behavior/output, that must be compensated for, and a resolution sought out unless the behavior is documented.

**Comment by fyrye** [comment]
issue is still pending review

**Comment by tobias-kuendig** [comment]
Thank you for the great additions @fyrye. It's a pity that this issue was closed by a bot since it is a problem that should be solved.

**Comment by FrancoisSt** [comment]
This is a serious issue for us. Will be implementing fix as proposed by @fyrye until resolved.

**Comment by fyrye** [comment]
Looking for a bit of feedback from the community before I submit a pull-request. There are a number of ways to implement a fix so that all of the date/time methods function identically, but it is a major BC break to any of the current applications relying on the feature. My preferred method is `#1` and using `setTimestamp()`

**Comment by ptrsz** [comment]
We have some issue that may relate to this one.

In `src/PhpSpreadsheet/Reader/Xml.php` at lines 531-535:
```
case 'DateTime':
                  $type = DataType::TYPE_NUMERIC;
                  $cellValue = Date::PHPToExcel(strtotime($cellValue));
   break;
```

The problem as we discovered is that the strtotime() function is working based on UTC timezone, and it does not take into consideration the date_default_timezone_get() setting.

Our solution was that we modified to above code as the following:
```
 $type = DataType::TYPE_NUMERIC; // original 

 $saveTimeZone = date_default_timezone_get(); // added
 date_default_timezone_set('UTC');  // added

 $cellValue = Date::PHPToExcel(strtotime($cellValue)); // original

 date_default_timezone_set($saveTimeZone); // added
```

**Comment by fyrye** [comment]
@ptrsz Correct, however `strtotime`  honors the PHP default timezone, the issue is caused by `Date::PHPToExcel()` using `new DateTime('@', . $dateValue)` when it receives an integer . Your solution will fix the issue only when using the Xml reader. 
I recommend changing [`src/PhpSpreadsheet/Shared/Date.php`](https://github.com/PHPOffice/PhpSpreadsheet/blob/master/src/PhpSpreadsheet/Shared/Date.php#L263) to ensure it functions the same through your application.

The generally accepted usage is use [`setTimestamp`](https://www.php.net/manual/en/datetime.settimestamp.php) and replace `new DateTime('@', . $dateValue)` with `(new DateTime())->setTimestamp($dateValue)`

**Comment by ptrsz** [comment]
@fyrye Thanks for the clarification, we will use your recommended solution method until the pull request is created and merged. Any -wild-  guesses when will that happen? (So we can revisit the issue accordingly)

**Comment by HalBert42** [comment]
Since this is php, can't in get the date.timezone = America/Los_Angeles from php ini?

**Comment by fyrye** [comment]
@HalBert42 `DateTime('@timestamp')` ignores the currently assigned timezone from `date.timezone` and any defined as the second argument of the `DateTime` constructor.  Source: https://www.php.net/manual/en/datetime.formats.compound.php 

> The "Unix Timestamp" format sets the timezone to UTC.

Example:  https://3v4l.org/jnfG5
```php
date_default_timezone_set('America/New_York');

$timestamp = time();

$at = new DateTime('@' . $timestamp);

$setTimestamp = (new DateTime())->setTimestamp($timestamp);

echo $at->format('r'); // //Sun, 16 Jun 2019 06:20:59 +0000
echo $setTimestamp->format('r'); //Sun, 16 Jun 2019 02:20:59 -0400
```

**Comment by HalBert42** [comment]
Fyrye, thank you for the reply,

 

Hal

**Comment by oleibman** [comment]
Doc block for `timestampToExcel` now states that using it is discouraged, and that it is not timezone-aware. Method `dateTimeFromTimestamp` is available and is timezone-aware, and its result can be used in `dateTimeToExcel`. Removing stale label, leaving issue closed.
